File: bin/Data/EditorLua/view/NodeTree.lua
local NodeTree = {}
NodeTree.isShow = true
local Translator = require("EditorLua/core/Translate")
require("EditorLua/core/BitUtils")

local OnClicked = function(node)
    NodeTree.app.cppApp:SelectNode(node)
end

local OnDoubleClicked = function()
end

local DrawNode
DrawNode = function(node)
    local nodeName = node:GetName()
    --ImGuiTreeNodeFlags_Selected
    --ImGuiTreeNodeFlags_OpenOnArrow | ImGuiTreeNodeFlags_OpenOnDoubleClick | ImGuiTreeNodeFlags_SpanAvailWidth;
    local flags = BitOr(imgui.constant.TreeNodeFlags.OpenOnArrow,imgui.constant.TreeNodeFlags.OpenOnDoubleClick)
    flags =  BitOr(flags,imgui.constant.TreeNodeFlags.SpanAvailWidth)
    local children = node:GetChildren(false)
    if NodeTree.app.cppApp:GetSelectNode() and node:GetID() == NodeTree.app.cppApp:GetSelectNode():GetID() then
        flags =  BitOr(flags,imgui.constant.TreeNodeFlags.Selected)
    end
    if(#children == 0) then
        --ImGuiTreeNodeFlags_Leaf |ImGuiTreeNodeFlags_NoTreePushOnOpen
        flags =  BitOr(flags,imgui.constant.TreeNodeFlags.Leaf)
    end
    local node_open = imgui.TreeNodeEx(nodeName,flags)
    --events
    if imgui.IsItemClicked(0) then
        OnClicked(node)
    end
    if imgui.IsMouseDoubleClicked(0) and imgui.IsItemHovered(0) then
        OnDoubleClicked(path)
    end
    --events end
    if node_open then
        --children
        for k,v in ipairs(children) do
            DrawNode(v)
        end
        imgui.TreePop()
    end
end

function NodeTree:Update()
    if self.isShow == true then
        self.isShow = imgui.Begin(Translator:translate("NodeTree"),self.isShow)
        DrawNode(self.app.cppApp:GetSceneRoot())
        if imgui.BeginPopupContextWindow("context",1) then
            imgui.MenuItem("create","",false,true)
            --imgui.BeginMenu sub menu
            imgui.EndPopup()
        end
        imgui.End()
    end
end

function NodeTree:Show(show)
    self.isShow = show
end

return NodeTree

File: Source/UrhoEditor/views/NodeTree.cpp
#include "NodeTree.h"
#include "EditorApp.h"
#include "Utils.h"

namespace Urho3DEditor 
{
NodeTree::NodeTree() 
{
}
NodeTree::~NodeTree() 
{
}
void NodeTree::Update() 
{
	if(!showing)
		return;
	ImGui::Begin("NodeTree",&showing);
	ImVec2 newSize = ImGui::GetWindowSize();
	_winPos = ImGui::GetWindowPos();
	if (newSize.x != winSize.x || newSize.y != winSize.y) {
		winSize = newSize;
	}
	DrawNode(EditorApp::GetInstance()->GetSceneRoot(),true);
	/*if(ImGui::BeginPopupContextWindow("context"),1)
	{
		ImGui::MenuItem("create","",false,true);
		ImGui::EndPopup();
	}*/
	ImGui::End();
}
void NodeTree::OnClicked(Node* node)
{
	EditorApp::GetInstance()->SelectNode(node);
}
void NodeTree::OnDoubleClicked() 
{
	
}

void NodeTree::DrawNodeNoInWindows(int itemH)
{
	//Draw nothing
	ImVec2 curPos = ImGui::GetCursorScreenPos();
	curPos.y = curPos.y + itemH;
	ImGui::SetCursorScreenPos(curPos);
}

void NodeTree::DrawNode(Node* node,bool isRoot)
{
	String nodeName = node->GetName();
	int flags = ImGuiTreeNodeFlags_OpenOnArrow | ImGuiTreeNodeFlags_OpenOnDoubleClick | ImGuiTreeNodeFlags_SpanAvailWidth;
	auto children = node->GetChildren(false);
	auto selecednode = EditorApp::GetInstance()->GetSelectNode();
	if(isRoot)
	{
		flags |= ImGuiTreeNodeFlags_DefaultOpen;
	}
	if(selecednode && node->GetID()== selecednode->GetID())
	{
		flags |= ImGuiTreeNodeFlags_Selected;
	}
	if(children.Size()== 0)
	{
		flags |= ImGuiTreeNodeFlags_Leaf;
	}
	if (IsInWindow(ImGui::GetCursorScreenPos()))
	{
		ImGui::PushID(node->GetID());
		bool node_open = ImGui::TreeNodeEx(nodeName.CString(), flags);
		ImGui::PopID();
		if (ImGui::IsMouseDoubleClicked(0) && ImGui::IsItemHovered()) {
			OnDoubleClicked();
		}
		else if (ImGui::IsItemClicked()) {
			OnClicked(node);
		}
		if (node_open) {
			if(children.Size() > 0)
			{
				_foldState[node->GetID()] = true;
			}
			for (auto item : children) {
				DrawNode(item,false);
			}
			ImGui::TreePop();
		}
	}
	else
	{
		DrawNodeNoInWindows(ImGui::GetItemRectSize().y);
		if (_foldState.Contains(node->GetID()) && _foldState[node->GetID()]) {
			for (auto item : children) {
				DrawNode(item,false);
			}
		}
	}
	
}
}